csrc/ir_utils.cpp

```diff
@@ -572,6 +574,83 @@ bool isReductionTvOp(const Expr* expr) {
   return ir_utils::isTvOp(expr) && isReductionOp(expr);
 }
 
+bool isReductionOverSizeZero(
+    const Expr* expr,
+    SchedulerRuntimeInfo& runtime_info) {
+  if (!isReductionOp(expr)) {
+    return false;
+  }
+  auto expr_eval = runtime_info.expressionEvaluator();
+  for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+    for (auto id : output->getRootDomain()) {
+      if (id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool reductionHasSizeZeroConcreteDimension(
+    const Expr* expr,
+    SchedulerRuntimeInfo& runtime_info) {
+  if (!isReductionOp(expr)) {
+    return false;
+  }
+  auto expr_eval = runtime_info.expressionEvaluator();
+  for (auto output : ir_utils::filterByType<TensorView>(expr->outputs())) {
+    for (auto id : output->getRootDomain()) {
+      if (!id->isReduction() &&
+          expr_eval.evaluate(id->extent())->as<int64_t>() == 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void replaceReductionWithFull(Expr* expr) {
+  auto fusion = expr->container()->as<Fusion>();
+  if (expr->isA<ReductionOp>()) {
+    // Fill value is reduction starting value
+    auto rop = expr->as<ReductionOp>();
+    auto old_out = rop->out()->as<TensorView>();
+    auto new_out = full_like(old_out, rop->init());
+    ir_utils::replaceValue(fusion, {{old_out, new_out}});
+    if (old_out->isFusionOutput()) {
+      fusion->replaceOutput(old_out, new_out);
+    }
+  } else if (expr->isA<WelfordOp>()) {
+    // Fill Welford avg/var with nan, N with 0
+    auto wop = expr->as<WelfordOp>();
```

Comment:
Yes I think it could. I originally had a little more logic in here that was wrong, so it was removed. Each output of a Welford is equal to the two reductions each divided by `N`. For size-0 reduction domain, this is 0/0. So we just need to add the division by 0 in the case of Welford, but otherwise it's the same logic as for any other reduction.